dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/statelisteners/ElasticListenerTest.java

```diff
@@ -0,0 +1,156 @@
+import io.dockstore.webservice.core.Tool;
+import io.dockstore.webservice.core.Version;
+import io.dockstore.webservice.core.WorkflowVersion;
+import java.util.List;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ElasticListenerTest {
+
+    private static final String FIRST_VERSION_NAME = "First";
+    private static final String SECOND_VERSION_NAME = "Second";
+    private static final int FIRST_VERSION_ID = 1;
+    private static final int SECOND_VERSION_ID = 2;
+    private WorkflowVersion firstWorkflowVersion;
+    private WorkflowVersion secondWorkflowVersion;
+    private Tag firstTag;
+    private Tag secondTag;
+    private WorkflowVersion firstAppToolVersion;
+    private WorkflowVersion secondAppToolVersion;
+    private BioWorkflow bioWorkflow;
+    private Tool tool;
+    private AppTool appTool;
+
+
+    @Before
+    public void setup() throws IllegalAccessException {
+
+        bioWorkflow = new BioWorkflow();
+        firstWorkflowVersion = new WorkflowVersion();
+        secondWorkflowVersion = new WorkflowVersion();
+        initVersion(firstWorkflowVersion, FIRST_VERSION_NAME, FIRST_VERSION_ID);
+        initVersion(secondWorkflowVersion, SECOND_VERSION_NAME,
+            SECOND_VERSION_ID);
+        bioWorkflow.getWorkflowVersions().addAll(List.of(firstWorkflowVersion,
+            secondWorkflowVersion));
+
+        tool = new Tool();
+        firstTag = new Tag();
+        secondTag = new Tag();
+        initVersion(firstTag, FIRST_VERSION_NAME, FIRST_VERSION_ID);
+        initVersion(secondTag, SECOND_VERSION_NAME, SECOND_VERSION_ID);
+        tool.getWorkflowVersions().addAll(List.of(firstTag, secondTag));
+
+        appTool = new AppTool();
+        firstAppToolVersion = new WorkflowVersion();
+        secondAppToolVersion = new WorkflowVersion();
+        initVersion(firstAppToolVersion, FIRST_VERSION_NAME, FIRST_VERSION_ID);
+        initVersion(secondAppToolVersion, SECOND_VERSION_NAME, SECOND_VERSION_ID);
+        appTool.getWorkflowVersions().addAll(List.of(firstAppToolVersion, secondAppToolVersion));
+    }
+
+    private void initVersion(final Version version, final String name, final long id)
+        throws IllegalAccessException {
+        version.setName(name);
+        final SourceFile sourceFile = new SourceFile();
+        sourceFile.setPath("/Dockstore.wdl");
+        sourceFile.setContent("Doesn't matter");
+        version.getSourceFiles().add(sourceFile);
+        // Id is normally set via Hibernate generator; have to use reflection to set it, alas
+        FieldUtils.writeField(version, "id", id, true);
+    }
+
+    @Test
+    public void testNoValidVersions() {
+        // If there are no valid versions, the latest version id wins out
+        List.of(bioWorkflow, tool, appTool).stream().forEach(entry -> {
+            final Entry detachedEntry = ElasticListener.removeIrrelevantProperties(entry);
```

Comment:
A general architectural commentary, no need to change anything:
This really is a test of the `removeIrrelevantProperties` method, rather than the version filtering capabilities of the `ElasticListener` class.  Sure, right now, they're sort of the same: `ElasticListener` uses `removeIrrelevantProperties` to do the filtering, but that could change (intentionally or otherwise), and this test would have no way of knowing.  If we were to abstract the search service, we could easily instantiate an `ElasticListener` that used a mocked service object, pass it an entry, and check that it emerged in the proper form at the service interface...